### PR TITLE
ARROW-14644: [C++][R] open_dataset doesn't ignore BOM in csv file

### PR DESCRIPTION
**Pull request by wjones127**
`arrow::csv::StreamingReader` already has handling for byte order marks (BOM). However, #7896 introduced `arrow::dataset::GetColumnNames` which is called prior to instantiating the reader and was missing BOM handling. This PR adds BOM handling to that method.

Without BOM handling, the first column as parsed by `arrow::dataset::GetColumnNames` contained the BOM (e.g. was `"<BOM>a"` instead of `"a"`). Because of this, it failed the test on line 120 below and was not added to `convert_options.include_columns`. 

https://github.com/apache/arrow/blob/9cf4275a19c994879172e5d3b03ade9a96a10721/cpp/src/arrow/dataset/file_csv.cc#L117-L122

### CODE REVIEW

File: cpp/src/arrow/dataset/file_csv.cc [hunk 1]
```diff
@@ -85,7 +86,13 @@ Result<std::unordered_set<std::string>> GetColumnNames(
 
   RETURN_NOT_OK(
       parser.VisitLastRow([&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
-        util::string_view view{reinterpret_cast<const char*>(data), size};
+        // Skip BOM when reading column names (ARROW-14644)
+        ARROW_ASSIGN_OR_RAISE(auto data2, util::SkipUTF8BOM(data, size));
```
Comment (westonpace) [inline cpp/src/arrow/dataset/file_csv.cc:90]: Minor nit: I don't love `data2` as a name.  Maybe `data_no_bom` or `sliced_data`.

File: cpp/src/arrow/dataset/file_csv.cc [hunk 2]
```diff
@@ -85,7 +86,13 @@ Result<std::unordered_set<std::string>> GetColumnNames(
 
   RETURN_NOT_OK(
       parser.VisitLastRow([&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
-        util::string_view view{reinterpret_cast<const char*>(data), size};
+        // Skip BOM when reading column names (ARROW-14644)
+        ARROW_ASSIGN_OR_RAISE(auto data_no_bom, util::SkipUTF8BOM(data, size));
+        int32_t offset = data_no_bom - data;
+        DCHECK_GE(offset, 0);
+        size = size - offset;
```
Comment (westonpace) [inline cpp/src/arrow/dataset/file_csv.cc:93]:
```suggestion
        ptrdiff_t offset = data_no_bom - data;
        DCHECK_GE(offset, 0);
        size = size - static_cast<uint32_t>(offset);
```
The Windows CI job is failing.

I think this will make windows happy but I never know and I don't have this in my IDE at the moment so I may be way off base here.

Comment (wjones127) [reply cpp/src/arrow/dataset/file_csv.cc]: Thanks, I wasn't sure what the best way to explicitly cast was. Those changes work locally on my Mac; hopefully will work in Windows builds.

File: r/tests/testthat/test-csv.R [hunk 3]
```diff
@@ -447,5 +447,13 @@ test_that("write_csv_arrow deals with duplication in include_headers/col_names",
   )
   expect_true(file.exists(csv_file))
   expect_identical(tbl_no_dates, written_tbl)
+})
+
+test_that("read_csv_arrow() deals with BOMs (bite-order-marks) correctly", {
+  writeLines('\xef\xbb\xbfa,b\n1,2\n', con = csv_file)
```
Comment (westonpace) [inline r/tests/testthat/test-csv.R:453]: Sorry, I should have noticed this before but it looks like there are some R style violations
```suggestion
  writeLines("\xef\xbb\xbfa,b\n1,2\n", con = csv_file)
```

File: r/tests/testthat/test-dataset-csv.R [hunk 4]
```diff
@@ -288,3 +289,12 @@ test_that("Column names inferred from schema for headerless CSVs (ARROW-14063)",
   ds <- open_dataset(headerless_csv_dir, format = "csv", schema = schema(int = int32(), dbl = float64()))
   expect_equal(ds %>% collect(), tbl)
 })
+
+test_that("open_dataset() deals with BOMs (bite-order-marks) correctly", {
+  writeLines('\xef\xbb\xbfa,b\n1,2\n', con = csv_file)
```
Comment (westonpace) [inline r/tests/testthat/test-dataset-csv.R:294]:
```suggestion
  writeLines("\xef\xbb\xbfa,b\n1,2\n", con = csv_file)
```

**Review by westonpace — APPROVED** [review]
Two last style nits and then I think this is good to go.  Thanks for getting to the bottom of this.

File: cpp/src/arrow/dataset/file_csv.cc [hunk 5]
```diff
@@ -85,7 +86,13 @@ Result<std::unordered_set<std::string>> GetColumnNames(
 
   RETURN_NOT_OK(
       parser.VisitLastRow([&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
-        util::string_view view{reinterpret_cast<const char*>(data), size};
+        // Skip BOM when reading column names (ARROW-14644)
+        ARROW_ASSIGN_OR_RAISE(auto data_no_bom, util::SkipUTF8BOM(data, size));
+        ptrdiff_t offset = data_no_bom - data;
+        DCHECK_GE(offset, 0);
```
Comment (pitrou) [inline cpp/src/arrow/dataset/file_csv.cc:92]: This doesn't seem useful. You can therefore shorten this to:
```c++
        size = size - static_cast<uint32_t>(data_no_bom - data);
```

File: r/tests/testthat/test-csv.R [hunk 6]
```diff
@@ -447,5 +447,13 @@ test_that("write_csv_arrow deals with duplication in include_headers/col_names",
   )
   expect_true(file.exists(csv_file))
   expect_identical(tbl_no_dates, written_tbl)
+})
+
+test_that("read_csv_arrow() deals with BOMs (bite-order-marks) correctly", {
```
Comment (pitrou) [inline r/tests/testthat/test-csv.R:452]: "byte"

File: r/tests/testthat/test-dataset-csv.R [hunk 7]
```diff
@@ -288,3 +289,12 @@ test_that("Column names inferred from schema for headerless CSVs (ARROW-14063)",
   ds <- open_dataset(headerless_csv_dir, format = "csv", schema = schema(int = int32(), dbl = float64()))
   expect_equal(ds %>% collect(), tbl)
 })
+
+test_that("open_dataset() deals with BOMs (bite-order-marks) correctly", {
```
Comment (pitrou) [inline r/tests/testthat/test-dataset-csv.R:293]: "byte"

**Review by pitrou — CHANGES_REQUESTED** [review]

**Review by pitrou — APPROVED** [review]
+1 for the fix. I'll let someone else comment on the R tests if needed.

File: r/tests/testthat/test-dataset-csv.R [hunk 8]
```diff
@@ -288,3 +289,12 @@ test_that("Column names inferred from schema for headerless CSVs (ARROW-14063)",
...
   expect_equal(ds %>% collect(), tbl)
 })
+
+test_that("open_dataset() deals with BOMs (byte-order-marks) correctly", {
+  writeLines("\xef\xbb\xbfa,b\n1,2\n", con = csv_file)
+
+  expect_equal(
+    open_dataset(csv_file, format = "csv") %>% collect(),
+    tibble(a = 1, b = 2)
+  )
```
Comment (jonkeane) [inline r/tests/testthat/test-dataset-csv.R:299]: Would it be possible to make two csv files and read those in as a dataset? This _should_ work just the same, but it would be good to confirm that the second csv has the proper BOM handling as well.

**Review by jonkeane — APPROVED** [review]
This is fantastic. It is fine to go, though I have one suggestion about exercising datasets a little bit more that you can take or leave.

File: r/tests/testthat/test-dataset-csv.R (hunk 8 above)
Comment (wjones127) [reply r/tests/testthat/test-dataset-csv.R:299]: Oh that's a good idea. I'll add that really quick.

Comment (dragosmg) [reply r/tests/testthat/test-dataset-csv.R:300]: @wjones127 @jonkeane I think the additional tests are now failing. I got [this](https://github.com/apache/arrow/runs/4470780115?check_suite_focus=true) in a different PR. I haven't yet rebased so I don't even have those tests in my branch.

Comment (pitrou) [reply r/tests/testthat/test-dataset-csv.R:300]: @dragosmg I filed https://issues.apache.org/jira/browse/ARROW-15041
